render-wasm/docs/tile_rendering.md

```diff
@@ -0,0 +1,202 @@
+
+When zooming or panning:
+
+- New tiles entering the `visible_rect` are identified.
+- These tiles are either:
+  - Pulled from the cache, or  
+  - Rendered fresh if not already available.
+
+### Handling Zoom and Scale
+
+- Tile resolution depends on zoom level (`scale`).
+- When zoomed in significantly, higher-resolution tiles are required.
+- If the zoom changes beyond a threshold, the cache may be invalidated, and tiles re-rendered at the new scale.
+
+## Why cache tiles?
+
+Tile caching (via `TileTextureCache`) boosts performance:
+
+- Prevents unnecessary rendering.
+- Only a limited number of tiles are kept in memory (`TEXTURES_CACHE_CAPACITY`).
+- Distant tiles are periodically evicted using an LRU-like strategy (based on visibility).
+
+## Scaling and Tile Size Utilities
+
+Tile size in screen space is constant, but visually it depends on zoom level (`scale`).
+
+To deal with this, the system uses utility functions:
+
+- `get_tile_size(scale)`: Gets the logical size of a tile at the current scale.
+- `get_tile_rect(tile, scale)`: Gets the rectangular area a tile covers.
+- `get_tile_pos(tile, scale)`: Gets the tile’s position in canvas coordinates.
+
+These allow rendering logic to adapt tile positions/sizes dynamically.
+
+## Selective Re-rendering
+
+The system enables:
+
+- Efficient updates when the viewport changes.
+- Targeted re-rendering when a shape changes (only the affected tiles).
+- Reduced GPU/CPU work on complex documents.
+
+## Rendering Pipeline Summary (Per Frame)
+
+1. **Input**: `document_state` + `Viewbox`  
+   - Shapes, layer order, transforms + visible area.
+
+2. **Compute `TileViewbox`**  
+   - From `Viewbox`, generate visible + interest regions.
+
+3. **Determine Tiles to Render**  
+   - Compare current tiles vs cached ones.
+   - Detect cache hits and misses.
+   - Invalidate stale tiles if viewport changed significantly.
+
+4. **Render From Cache**  
+   - For cached tiles:
+     - Fetch image from `TileTextureCache`.
+     - Blit to `Surface::Target`.
+
+5. **Render New Tiles**  
+   - For non-cached tiles:
+     - Find intersecting shapes (`TileHashMap`).
+     - Render layers in this order:
+       - Fills → Strokes → Shadows → Composite → Target
+     - Each layer is rendered on a `skia::Surface`.
+     - Use blending, opacity, masking, etc.
+     - Store the result in the cache:  
+       `cache_current_tile_texture(tile, image)`
+
+6. **Draw to Screen**  
+   - Once a tile is rendered, draw it to `Surface::Target`.
+
+7. **Handle Interactions**  
+   - When shapes are modified:
+     - Determine affected tiles.
+     - Invalidate those tiles in the cache.
+     - Re-render them selectively.
+   - When panning or zooming:
+     - Detect new visible tiles.
+     - Use cache or render them if needed.
+
+## Summary of Benefits
+
+- **Modular pipeline**: Separate surfaces per layer enable composable rendering.
+- **Cache-friendly**: Avoids re-rendering when nothing has changed.
```

Comment:
IMHO the cache is the how, not the benefit. I would say: 

**Avoid unnecessary re-renders**: A cache is used to avoid re-rendering when nothing has changed.